### DISCUSSION TITLE
Storage limits

### DISCUSSION THREAD

**Discussion by Serega007RU** [tsukasa/BdBrowser#3] [General]
What about storage limits? 5 mb is very small, is there a way to increase this limit?
![image](https://user-images.githubusercontent.com/64780091/187124775-72bebe1c-11e8-4961-bd3f-3cecfd77b85a.png)

**Comment by tsukasa** [comment]
Greetings,

SessionStorage/LocalStorage has a hard limit of 5MB[1].

I am currently experimenting with a very rough new version of the fs module that uses IndexedDB[2] as its backend, which should provide three key advantages over the current solution:
1. Up to 1GB of storage (per specification).
2. It can save Blobs, so serialization/deserialization should no longer be an issue (in terms of CPU usage or supported file types).
3. Easier administration (from an end-user perspective, allows to locate/edit/remove records via the Chrome developer tools).

Unfortunately, it is not all roses and sunshine though. IndexedDB is asynchronous with no synchronous API available, meaning that synchronous functions like unlinkSync() need to be rebuilt with some sort of lock that wastes CPU time waiting for the promise to be resolved to emulate synchronous operation for the function (i.e. by looping as long as the promise is not resolved, with an Atomics-based wait of 10-50ms between loops; better suggestions are welcome!).

The synchronous fs functions are required for BetterDiscord to work, so skipping them is not an option.

I cannot give any concrete timeframe for when this work will hit the repository, as it is still highly in flux. But it is actively being worked on.

References:
[1] [Storage for the web](https://web.dev/storage-for-the-web)
[2] [IndexedDB API](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API)

**Comment by tsukasa** [comment]
Small update for those interested/subscribed to this topic and the state of this task.

### The Bad
Despite JavaScript being a single-threaded language, there is apparently no way to deliberately block execution programmatically like in other programming languages. This sounds good until you notice that there is also no waitAsync or processEvents equivalent and that the end of a loop run does not trigger the event processing. Since the `fs` module API is a fixed target to serve, following the public sentiment of "just adapt your design and use Promises" is impossible for this specific use-case.

As mentioned before, IndexedDB is asynchronous and, as I recently learned, so are the operations to convert `Blob` objects. They either require callbacks or return promises. This means there are a few design alterations for this task:

1. Blobs are no longer being used/targeted for this task.
   Back to text-only storage via strings. Who wants non-text data in their BetterDiscord directory anyway? Unfortunately `Blob.text()` returns promises and I cannot process these in the existing program flow. Either because I have not fully understood them (lack of knowledge) or because I have understood them and JavaScript does not offer the language features for this use-case.

2. Direct IndexedDB usage is not possible.
   I need to push a layer in between the application and the vfs data store, same story as with `Blob`. This layer is an in-memory cache that gets flood filled during site load and will enable the use of the synchronous `fs` functions BetterDiscord requires while still being able to persist data in IndexedDB in the background asynchronously. This comes at the cost of RAM because an entire copy of the VFS needs to be held in memory for the synchronous functions to work.

### The Good
Do not let the previous points turn you off (except if you rock only 512MB RAM!).

1. The 5MB limit is gone.
   This is useful for people that use a certain plugin to log messages, I imagine. Or folks who always wanted to download the 5MB German dictionary for the Spellchecker plugin. The point is: You can use more than 5MB of space with the new system.

2. Exploring the VFS and deleting entries just got a lot easier.
   You can use the IndexedDB browser in the Chrome developer tools to explore what is in your VFS and delete entries from there (requires site reload afterwards to sync the in-memory cache).

3. The memory costs are not out of this world.
   Considering Discord itself already gobbles up 500+MB of RAM, another 5-20MB (depending on your VFS size) should not be such a big deal. If push comes to shove, data lands in the swap file. I do not really see this as an issue in 2022 anymore, especially since the target audience for this extension probably already has somewhat capable machines.

4. All of your existing VFS files will be migrated.
   No need to download/configure plugins and themes after the automatic migration. Ideally, you will not even notice that you use a new VFS system. Fun fact: Migration takes about 60ms for me.

5. The old VFS stays in the house.
   Should you decide that you want to go back and fork the repository because this VFS solution is not for you: Good news, the bd-files stay untouched. This is also good news in case the migration goes wrong or you erase the database by mistake.

6. inotify got improved.
   The hacky solution to dispatch the "change" and "rename" events to filesystem watchers is no longer necessary and has been replaced with a proper solution that will also work its way up the filesystem to inform the parent elements.

### The Rest
The following changes are of note for the upcoming release:

1. unlink quirks.
   Unlink will be able to delete folders. Deleting a folder will recursively delete all of its contents as well. Yes, this is against specification and will later be split into the proper unlink, rmSync and rmdir functions. Testing the current set of changes takes priority over refactoring this.

2. statSync is more feature complete.
   Calling statSync will allow to read properties like the birthtime. Access time is present, however does not get updated yet.

3. It is easy to roll back to the old VFS system in the code.
   If you want to roll back to the old LocalStorage-based system, you can simply replace every reference to fs.js with fs-localstorage.js in the code and webpack the project. The old file will stay part of the project.

### Release Schedule
I am currently field-testing the feature-complete package before pushing the commit to the master branch, so that stability issues, API incompatibilities/quirks, performance issues and onboarding/migration problems can be taken care of beforehand to make this something users can look forward to.

In case everything goes smoothly, I anticipate that I will push the changes next weekend (somewhere in the vicinity of 2022-09-10).

**Reply by tsukasa** [reply]
From the `net::ERR_CONNECTION_RESET` error in your console, I would guess there are temporary issues on the side of GitHub's CDN. I base this on the `SupporterBadge.css` - it gets fetched directly (without any shimming involved) and still fails. It is likely the other files fail with a similar message, however that message stays in the service worker's log (accessible through Chrome's Extension page).

Nothing in the last 4 VFS commits has touched anything related to the request and fetch handling.

**Comment by Serega007RU** [comment]
Wow, that's really cool! Discord is now much faster to launch (especially when you have 50+ plugins). Thank you very much for the work done, it has become much more comfortable to work in web discord.
